### PR TITLE
Use BytePointer wherever possible

### PR DESCRIPTION
**Pull request by EmergentOrder**
Remove usage of ByteBuffer, nioBuffer(), asBuffer(), DirectBuffer
Also bump Scala and sbt versions

### CODE REVIEW

**Comment by EmergentOrder** [comment]
FYI, In this PR I also removed 2 usages of `Integer.reverseBytes` which accounted for the endianness of ByteBuffer vs BytePointer. Given that the last PR removed calls like `v_bb.order(ByteOrder.LITTLE_ENDIAN)` , there could be a gap there which is resolved with this PR.

**Comment by wmeddie** [comment]
I run into this when running it on yarn.

```
22/02/04 10:19:19 INFO YarnClusterScheduler: Adding task set 0.0 with 6 tasks resource profile 0
22/02/04 10:19:34 WARN YarnClusterScheduler: Initial job has not accepted any resources; check your cluster UI to ensure that workers are registered and have sufficient resources
22/02/04 10:19:49 WARN YarnClusterScheduler: Initial job has not accepted any resources; check your cluster UI to ensure that workers are registered and have sufficient resources
22/02/04 10:20:04 WARN YarnClusterScheduler: Initial job has not accepted any resources; check your cluster UI to ensure that workers are registered and have sufficient resources
22/02/04 10:20:19 WARN YarnClusterScheduler: Initial job has not accepted any resources; check your cluster UI to ensure that workers are registered and have sufficient resources
22/02/04 10:20:34 WARN YarnClusterScheduler: Initial job has not accepted any resources; check your cluster UI to ensure that workers are registered and have sufficient resources
22/02/04 10:20:48 INFO YarnAllocator: Completed container container_1643543749702_0051_01_000002 on host: XAIJPVE1 (state: COMPLETE, exit status: 1)
22/02/04 10:20:48 WARN YarnAllocator: Container from a bad node: container_1643543749702_0051_01_000002 on host: XAIJPVE1. Exit status: 1. Diagnostics: [2022-02-04 10:20:47.610]Exception from container-launch.
Container id: container_1643543749702_0051_01_000002
Exit code: 1

[2022-02-04 10:20:47.611]Container exited with a non-zero exit code 1. Error file: prelaunch.err.
Last 4096 bytes of prelaunch.err :
Last 4096 bytes of stderr :
ty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at org.apache.spark.network.util.TransportFrameDecoder.channelRead(TransportFrameDecoder.java:102)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Exception in thread "main" java.lang.reflect.UndeclaredThrowableException
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1748)
	at org.apache.spark.deploy.SparkHadoopUtil.runAsSparkUser(SparkHadoopUtil.scala:61)
	at org.apache.spark.executor.CoarseGrainedExecutorBackend$.run(CoarseGrainedExecutorBackend.scala:393)
	at org.apache.spark.executor.YarnCoarseGrainedExecutorBackend$.main(YarnCoarseGrainedExecutorBackend.scala:81)
	at org.apache.spark.executor.YarnCoarseGrainedExecutorBackend.main(YarnCoarseGrainedExecutorBackend.scala)
Caused by: org.apache.spark.rpc.RpcTimeoutException: Futures timed out after [120 seconds]. This timeout is controlled by spark.rpc.askTimeout
	at org.apache.spark.rpc.RpcTimeout.org$apache$spark$rpc$RpcTimeout$$createRpcTimeoutException(RpcTimeout.scala:47)
	at org.apache.spark.rpc.RpcTimeout$$anonfun$addMessageIfTimeout$1.applyOrElse(RpcTimeout.scala:62)
	at org.apache.spark.rpc.RpcTimeout$$anonfun$addMessageIfTimeout$1.applyOrElse(RpcTimeout.scala:58)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38)
	at org.apache.spark.rpc.RpcTimeout.awaitResult(RpcTimeout.scala:76)
	at org.apache.spark.rpc.RpcEndpointRef.askSync(RpcEndpointRef.scala:103)
	at org.apache.spark.rpc.RpcEndpointRef.askSync(RpcEndpointRef.scala:87)
	at org.apache.spark.executor.CoarseGrainedExecutorBackend$.$anonfun$run$7(CoarseGrainedExecutorBackend.scala:421)
	at org.apache.spark.deploy.SparkHadoopUtil$$anon$1.run(SparkHadoopUtil.scala:62)
	at org.apache.spark.deploy.SparkHadoopUtil$$anon$1.run(SparkHadoopUtil.scala:61)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1730)
	... 4 more
Caused by: java.util.concurrent.TimeoutException: Futures timed out after [120 seconds]
	at scala.concurrent.impl.Promise$DefaultPromise.ready(Promise.scala:259)
	at scala.concurrent.impl.Promise$DefaultPromise.result(Promise.scala:263)
	at org.apache.spark.util.ThreadUtils$.awaitResult(ThreadUtils.scala:293)
	at org.apache.spark.rpc.RpcTimeout.awaitResult(RpcTimeout.scala:75)
	... 12 more

```

File: src/main/scala/com/nec/arrow/ArrowInterfaces.scala [hunk 1]
```diff
@@ -43,65 +40,58 @@ object ArrowInterfaces {
...
     vc
   }
 
   def c_bounded_string(string: String): non_null_c_bounded_string = {
     val vc = new non_null_c_bounded_string()
-    vc.data =  (new BytePointer(string.length)).asBuffer
-      .put(string.getBytes())
-      .asInstanceOf[DirectBuffer]
+    vc.data =  (new BytePointer(string.length))
+      .put(string.getBytes(): _*)
```
Comment (saudet) [inline src/main/scala/com/nec/arrow/ArrowInterfaces.scala:71]: Are we sure this doesn't need to be null-terminated?

File: src/main/scala/com/nec/ve/colvector/VeColVector.scala [hunk 2]
```diff
@@ -145,17 +143,17 @@ final case class VeColVector(underlying: GenericColVector[Long]) {
       if (numItems > 0) {
         val dataSize = numItems * 8
         float8Vector.setValueCount(numItems)
-        val vhTarget = ByteBuffer.allocateDirect(dataSize)
-        val validityTarget = ByteBuffer.allocateDirect(numItems)
+        val vhTarget = new BytePointer(ByteBuffer.allocateDirect(dataSize))
+        val validityTarget = new BytePointer(ByteBuffer.allocateDirect(numItems))
```
Comment (saudet) [inline src/main/scala/com/nec/ve/colvector/VeColVector.scala:147]: We don't need to call those allocateDirect() here or anywhere else. Use BytePointer to allocate memory.

Comment (EmergentOrder) [reply src/main/scala/com/nec/ve/colvector/VeColVector.scala:147]: Right, missed those, will remove.

File: src/main/scala/com/nec/arrow/ArrowInterfaces.scala (hunk 1 above)
Comment (saudet) [reply src/main/scala/com/nec/arrow/ArrowInterfaces.scala:71]: Right, it stores the length separately, so I guess this is alright, but I don't even see where the data gets used!

**Comment by wgip** [comment]
Getting this in `main` at the moment, when running `~ VectorEngine / testOnly *ExchangeOnLocalSpec*`, I think was some other commit that was integrated earlier

```
[info]   Cause: java.lang.ArrayIndexOutOfBoundsException:
[info]  at java.lang.System.arraycopy(Native Method)
[info]  at com.nec.arrow.colvector.ByteArrayColVector.$anonfun$serialize$4(ByteArrayColVector.scala:64)
[info]  at com.nec.arrow.colvector.ByteArrayColVector.$anonfun$serialize$4$adapted(ByteArrayColVector.scala:63)
[info]  at scala.collection.immutable.List.foreach(List.scala:392)
[info]  at com.nec.arrow.colvector.ByteArrayColVector.serialize(ByteArrayColVector.scala:63)
[info]  at com.nec.ve.colvector.VeColVector.$anonfun$serialize$1(VeColVector.scala:62)
[info]  at org.apache.spark.metrics.source.ProcessExecutorMetrics.measureRunningTime(ProcessExecutorMetrics.scala:55)
[info]  at com.nec.ve.colvector.VeColVector.serialize(VeColVector.scala:65)
[info]  at com.nec.ve.VeRDD$.$anonfun$exchange$1(VeRDD.scala:20)
[info]  at scala.collection.Iterator$$anon$10.next(Iterator.scala:459)
[info]  at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:156)
```

**Comment by EmergentOrder** [comment]
@wgip Thanks for reporting that, I can reproduce that test failure from the last commit from my last PR.

I just ran the same test with the branch for this PR, and it now passes.

So it will be resolved when this is merged, no need to revert it as you did here: https://github.com/XpressAI/SparkCyclone/pull/479

**Comment by EmergentOrder** [comment]
@wmeddie I can't reproduce the rpc timeout. 
Instead I get an `java.lang.UnupportedOperationException` on query 22, which I believe is unrelated to this PR.

**Comment by wgip** [comment]
Thanks @EmergentOrder for quick turnaround! Will integrate a little later.

**Comment by EmergentOrder** [comment]
Closed by https://github.com/XpressAI/SparkCyclone/pull/480

**Pull request closed without merging**